### PR TITLE
[tests] expand flex-attn test for vision models

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

As per title, skips the test unless all sub-models support flex attn. Some models didn't have the flag set even though they can support Flex attn, this PR goes over all recent;y refactored vision models and sets the flags to `True`. Flex attention tests are passing for all models (except Zamba2 fixed in the linked PR below) 

We can merge this to make CI green in VLMs for now, and as a long-term solution would be nice to remove these flags. Or at least to check only on the base model, instead of going into each backbone

### CODE REVIEW

**Review by vasqu — COMMENTED** [review]
Just a general question / concern: We only manipulate the text backbones. Are vision parts of the model that use attention not affected?

Imo, we have to be careful in vision as well that the hidden dims add up. no? (Atm it seems most vlms have a high enough dim that we do not need this)

**Comment by zucchini-nlp** [comment]
Hmm, I guess that is the reason tests didn't fail. Also, most vision-only models aren't flagged to support flex yet, so the test gets skipped for llava-like models.

I will add a similar guard on vision hidden sizes as it doesn't hurt, thanks

File: src/transformers/models/colpali/modeling_colpali.py [hunk 1]
```diff
@@ -162,7 +162,8 @@ def forward(
         # L2 normalization
         embeddings = embeddings / embeddings.norm(dim=-1, keepdim=True)  # (batch_size, sequence_length, dim)
 
-        embeddings = embeddings * attention_mask.unsqueeze(-1)  # (batch_size, sequence_length, dim)
+        if attention_mask is not None:
+            embeddings = embeddings * attention_mask.unsqueeze(-1)  # (batch_size, sequence_length, dim)
```
Comment (ydshieh) [inline src/transformers/models/colpali/modeling_colpali.py:166]: Make sense to me. And this is the only place I find when I search

> embeddings = embeddings * attention_mask.unsqueeze(-1)

and in `examples/modular-transformers/modular_new_task_model.py`.

Let's apply the same change there too?

File: tests/models/musicgen/test_modeling_musicgen.py [hunk 2]
```diff
@@ -569,6 +569,7 @@ class MusicgenTest(ModelTesterMixin, GenerationTesterMixin, PipelineTesterMixin,
     all_generative_model_classes = ()
     greedy_sample_model_classes = (MusicgenForConditionalGeneration,) if is_torch_available() else ()
     pipeline_model_mapping = {"text-to-audio": MusicgenForConditionalGeneration} if is_torch_available() else {}
+    additional_model_inputs = ["decoder_input_ids"]
```
Comment (ydshieh) [inline tests/models/musicgen/test_modeling_musicgen.py:573]: Maybe a short comment why such change here?

Is it some tests failing without this? If so, which tests? (flex ones?)

File: tests/models/siglip/test_modeling_siglip.py [hunk 3]
```diff
@@ -103,7 +103,7 @@ def __init__(
         patch_size=2,
         num_channels=3,
         is_training=True,
-        hidden_size=32,
+        hidden_size=64,
```
Comment (ydshieh) [inline tests/models/siglip/test_modeling_siglip.py:106]: short comment would be nice 🙏 (I mean in PR page, not directly in code)

File: tests/models/videomae/test_modeling_videomae.py [hunk 4]
```diff
@@ -190,7 +190,7 @@ class VideoMAEModelTest(ModelTesterMixin, PipelineTesterMixin, unittest.TestCase
         if is_torch_available()
         else {}
     )
-
+    additional_model_inputs = ["bool_masked_pos"]
```
Comment (ydshieh) [inline tests/models/videomae/test_modeling_videomae.py:194]: same comment 🙏

File: tests/test_modeling_common.py [hunk 5]
```diff
@@ -3638,7 +3638,8 @@ def test_eager_matches_sdpa_inference(
                 processed_inputs[model.main_input_name] = inputs_dict[model.main_input_name]
 
                 for key in getattr(self, "additional_model_inputs", []):
-                    processed_inputs[key] = inputs_dict[key]
+                    if key in inputs_dict:
+                        processed_inputs[key] = inputs_dict[key]
```
Comment (ydshieh) [inline tests/test_modeling_common.py:3643]: I am sure you did this with a valid reason.

However, it looks a bit strange itself: why we add something to `additional_model_inputs` if it would not be in `inputs_dict`?

Would like to get a bit of context here.

File: tests/test_modeling_common.py [hunk 6]
```diff
@@ -4013,19 +4014,21 @@ def test_flash_attn_2_can_dispatch_composite_models(self):
                 model = model_class.from_pretrained(tmpdirname, torch_dtype=torch_dtype)
 
                 sub_models_supporting_fa2 = [
-                    (module._supports_flash_attn_2 or module._supports_attention_backend)
```
Comment (ydshieh) [inline tests/test_modeling_common.py:4015]: Actually, I am behind the latest fast peace development 😭 . If you have time to help me, could you explain a bit what `_supports_attention_backend` does (I read the doc).

For example, `blip_2` has `attention_interface: Callable = eager_attention_forward` but `Blip2QFormerModel` has `_supports_attention_backend = False` 🤔

File: tests/test_modeling_common.py [hunk 7]
```diff
@@ -4569,33 +4572,67 @@ def recursively_check(eager_outputs, exported_outputs):
...
+                if hasattr(config, "head_dim") and config.head_dim is not None:
+                    head_dim = config.head_dim
+                    config.head_dim = max(16, config.head_dim)
+
+                if (
+                    getattr(config, "hidden_size", None) is not None
+                    and getattr(config, "num_attention_heads", None) is not None
+                ):
+                    head_dim = config.hidden_size // config.num_attention_heads
+                    config.hidden_size *= max(16 // head_dim, 1)
```
Comment (ydshieh) [inline tests/test_modeling_common.py]: if `config` has `head_dim`, we can use it. Only if it does not have this attr., we have to compute as `config.hidden_size // config.num_attention_heads`, right? So the condition could be relaxed ?

File: tests/test_modeling_common.py [hunk 8]
```diff
@@ -4569,33 +4572,67 @@ def recursively_check(eager_outputs, exported_outputs):
...
             self.assertTrue(model.config._attn_implementation == "flex_attention")
 
             # Elaborate workaround for encoder-decoder models as some do not specify their main input
             dummy_inputs = {model.main_input_name: inputs_dict[model.main_input_name].to(torch_device)}
-            if config.is_encoder_decoder:
+            for key in getattr(self, "additional_model_inputs", []):
+                if key in inputs_dict:
+                    dummy_inputs[key] = inputs_dict[key].to(torch_device)
+
+            if config.get_text_config(decoder=True).is_encoder_decoder:
```
Comment (ydshieh) [inline tests/test_modeling_common.py:4646]: Just to be sure : is this only relevant to models with `text_config` that is `encoder_decoder`, or it is relevant to models that doesn't have `text_config` and itself is `encoder_decoder`?

**Review by ydshieh — APPROVED** [review]
LGTM with a few nits comments/question.

Thank you for taking care of this 🙏

File: tests/models/musicgen/test_modeling_musicgen.py (hunk 2 above)
Comment (zucchini-nlp) [reply tests/models/musicgen/test_modeling_musicgen.py:573]: Yeah, this is the class attribute added by SDPA tests and we can re-use it in flex attention as well. I will add comment in the Mixin Test

The `additional_model_inputs` is used to specify input keys that are required for mode's forward. In music gen `input ids` isn't enough to run a forward pass even though the model isn't encoder-decoder in its config

File: tests/test_modeling_common.py (hunk 5 above)
Comment (zucchini-nlp) [reply tests/test_modeling_common.py:3643]: There;s one model in `all_model_classes` that has a different set of prepared input kwargs for forward, so we can test the model under different settings. We check if the `key` is present in the input dict or not therefore

I will add this comment in code :)

File: tests/test_modeling_common.py (hunk 6 above)
Comment (zucchini-nlp) [reply tests/test_modeling_common.py:4015]: That is my bad, it wasn't supposed to be merged in the first place...

The `_supports_attention_backend` is used by external developers like vLLM so they can set and use their own attn implementation, with optimized kernels. They just have to "register" the attention in advance and call `attn_implementation="vllm"`. Basically all models refactored for new attention interface support attention backend

I made this change as a pre-cursor for refactoring so we can get rid of tons of `_suppots-XXX` flags and didn't have time to get back. So we can't yet assume that any model with `_supports_attention_backend` support FA2/SDPA/Flex/anything, because some models can't support due to their architecture, even if we can pass the attention downstream

File: tests/test_modeling_common.py (hunk 7 above)
Comment (zucchini-nlp) [reply tests/test_modeling_common.py]: Yeah, I can update that. I made this for sub-configs which don't have any attention-related keys because this helper fn is applied recursively to all configs in the model now

File: tests/test_modeling_common.py (hunk 8 above)
Comment (zucchini-nlp) [reply tests/test_modeling_common.py:4646]: Both, if the model is LLM and has no `text_config` then `config.get_text_config(decoder=True)` returns the `self`

File: tests/models/siglip/test_modeling_siglip.py (hunk 3 above)
Comment (zucchini-nlp) [reply tests/models/siglip/test_modeling_siglip.py:106]: Siglip sets `config.projection` at init time to `config.hidden_size`, so even if we update config values in the test case it won't work and fail. Instead of adding more code only for Siglip, I decided to update the Tester

File: examples/modular-transformers/modeling_new_task_model.py
Comment (ydshieh) [inline examples/modular-transformers/modeling_new_task_model.py:1]: wow, didn't know it will be a huge change here 😅

Comment (zucchini-nlp) [reply examples/modular-transformers/modeling_new_task_model.py:1]: hehe, seems like the modular wasn't converted again since created

**Comment by zucchini-nlp** [comment]
Phew, messed up branches when working. Blindly force pushing isn't good
